xdsl/interpreters/linalg.py

```diff
@@ -61,3 +63,98 @@ def run_yield(
         self, interpreter: Interpreter, op: linalg.YieldOp, args: tuple[Any, ...]
     ):
         return ReturnedValues(args), ()
+
+    @impl(linalg.AddOp)
+    def run_add(
+        self, interpreter: Interpreter, op: linalg.AddOp, args: tuple[Any, ...]
+    ) -> tuple[Any, ...]:
+        lhs, rhs = args[0], args[1]
+        assert isinstance(lhs, ShapedArray)
+        assert isinstance(rhs, ShapedArray)
+        lhs = cast(ShapedArray[float], lhs)
+        rhs = cast(ShapedArray[float], rhs)
+        assert lhs.shape == rhs.shape
+        return (
+            ShapedArray(list(l + r for l, r in zip(lhs.data, rhs.data)), lhs.shape),
+        )
+
+    @impl(linalg.FillOp)
+    def run_fill(
+        self, interpreter: Interpreter, op: linalg.FillOp, args: tuple[Any, ...]
+    ) -> tuple[Any, ...]:
+        (operand,) = args
+        assert isinstance(operand, ShapedArray)
+        operand = cast(ShapedArray[float], operand)
+        result_type = op.res[0].type
+        assert isinstance(result_type, TensorType)
+        result_shape = list(result_type.get_shape())
+        return (
+            ShapedArray(list(operand.data * math.prod(result_shape)), result_shape),
+        )
+
+    @impl(linalg.MulOp)
+    def run_mul(
+        self, interpreter: Interpreter, op: linalg.MulOp, args: tuple[Any, ...]
+    ) -> tuple[Any, ...]:
+        lhs, rhs = args[0], args[1]
+        assert isinstance(lhs, ShapedArray)
+        assert isinstance(rhs, ShapedArray)
+        lhs = cast(ShapedArray[float], lhs)
+        rhs = cast(ShapedArray[float], rhs)
+        assert lhs.shape == rhs.shape
+        return (
+            ShapedArray(list(l * r for l, r in zip(lhs.data, rhs.data)), lhs.shape),
+        )
```

Comment:
we want to modify the out parameter in place, and return it

Reply:
to clarify, you mean, do in place addition the the lhs arg and return it in the shaped array,
so return` shapedarray(lhs.data, lhs.shape)`

Reply:
My understanding of the linalg args is that they pass the destination as the last parameter, in this case the arg, which should be modified. At the `tensor` level, this is sort of implicit, but at the `memref` level the result is not returned, just modified in-place, I think.

Reply:
I think we need to add the functionality to check the number of operands and arguments to the interpreter. If they're not the same count we have a problem.

Reply:
If you could open a separate PR for this that would be very helpful

Reply:
> I think we need to add the functionality to check the number of operands and arguments to the interpreter. If they're not the same count we have a problem.

this would be modifying `run_op` correct?

Reply:
> My understanding of the linalg args is that they pass the destination as the last parameter, in this case the arg, which should be modified. At the `tensor` level, this is sort of implicit, but at the `memref` level the result is not returned, just modified in-place, I think.

oh so modify to the expected result and just `return ()`

Reply:
I'm not sure how much of this context you're familiar with, so I'll describe my understanding of the whole thing:
In MLIR, there are two high-level shaped pointer abstractions, tensor and memref. Tensor is a value type, meaning that it's conceptually never mutated, and the resulting tensor is returned instead, so you can pretend that it's like a normal SSA value. Memref is a reference type, meaning that operations mutate what the memref contains. There's a process called bufferization that converts between the two. When bufferizing linalg, operations that operate on tensors are transformed into operations on memrefs, and in the process ops that return a value are replaced with ops that do not. So the operation implementation in the interpreter needs to check either the number of results or the types to check what the behaviour should be to be consistent with the behaviour both before and after bufferization.

Reply:
i feel like i understand, but ill reclarify: tensors aren't mutable but given memrefs are mutuable the operation value is kept somewhere but not returned explicitly like tensors, so to ensure we are consistent we must check that the no of results from tensor -> memref are the same  before /after we modify the args in place

Reply:
if you could include a general pseudocode skeleton of what you mean that be helpful too :)

Reply:
Basically all these functions need to mutate the out parameters in place, and return the out parameters if the op has results

Reply:
okay cool